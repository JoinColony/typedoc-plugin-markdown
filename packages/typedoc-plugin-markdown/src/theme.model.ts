import { PageEvent, ReflectionKind, RenderTemplate } from 'typedoc';

export interface TypedocPluginMarkdownOptions {
  hideBreadcrumbs: boolean;
  hideInPageTOC: boolean;
  hidePageTitle: boolean;
  hideMembersSymbol: boolean;
  entryDocument: string;
  entryPoints: string[];
  includes: string;
  indexTitle: string;
  media: string;
  namedAnchors: boolean;
  readme: string;
  publicPath: string;
}

export interface TemplateMapping {
  /**
   * The reflection this rule applies to.
   */
  kind: ReflectionKind[];

  /**
   * Can this mapping have children or should all further reflections be rendered
   * to the defined output page?
   */
  isLeaf: boolean;

  /**
   * The name of the directory the output files should be written to.
   */
  directory: string;

  /**
   * The name of the template that should be used to render the reflection.
   */
  template: RenderTemplate<PageEvent<any>>;
}
