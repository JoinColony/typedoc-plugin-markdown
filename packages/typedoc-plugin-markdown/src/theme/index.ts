export * from './theme';
export * from './theme.context';
export * from './theme.model';
