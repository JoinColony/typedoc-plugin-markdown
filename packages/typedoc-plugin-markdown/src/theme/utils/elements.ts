export class MarkdownString {
  lines: string[] = [];
  markdownString = '';

  ln(md: string) {
    this.markdownString = this.markdownString + '\n\n';
    this.lines.push(md);
  }

  space(md: string) {
    this.lines.push(md);
  }

  render(join: 'br' | 'space' = 'br') {
    return this.lines.join(join === 'br' ? '\n\n' : ' ');
  }

  render2() {
    return this.markdownString;
  }

  heading(level: number, text: string) {
    `${[...Array(level)].map(() => '#').join('')} ${text}`;
  }
}

export const heading = (level: number, text: string) =>
  `${[...Array(level)].map(() => '#').join('')} ${text}`;

export const link = (label: string, url: string) => `[${label}](${url})`;

export const bold = (text: string) => `**${text}**`;

export const backTicks = (text: string) => `\`${text}\``;

export const unorderedList = <T>(items: T[]) =>
  items.map((item) => `- ${item}`).join('\n');

export const horizontalRule = () => '---';

export const table = (headers: string[], rows: string[][]) =>
  `\n| ${headers.join(' | ')} |\n| ${headers
    .map(() => ':------')
    .join(' | ')} |\n${rows.map((row) => `| ${row.join(' | ')} |\n`).join('')}`;
