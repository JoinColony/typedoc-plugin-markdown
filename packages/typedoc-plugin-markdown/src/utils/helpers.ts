import { ReflectionKind } from 'typedoc';
import { FrontMatterVars } from './models';

export const KIND_PLURALS = {
  [ReflectionKind.Class]: 'Classes',
  [ReflectionKind.Property]: 'Properties',
  [ReflectionKind.Enum]: 'Enumerations',
  [ReflectionKind.EnumMember]: 'Enumeration embers',
  [ReflectionKind.TypeAlias]: 'Type Aliases',
};

export function getKindPlural(kind: ReflectionKind): string {
  if (kind in KIND_PLURALS) {
    return KIND_PLURALS[kind as keyof typeof KIND_PLURALS];
  } else {
    return getKindString(kind) + 's';
  }
}

export function getKindString(kind: ReflectionKind): string {
  let str = ReflectionKind[kind];
  str = str.replace(/(.)([A-Z])/g, (_m, a, b) => a + ' ' + b.toLowerCase());
  return str;
}

/**
 * Prepends YAML block to a string
 * @param contents - the string to prepend
 * @param vars - object of required front matter variables
 */
export const prependYAML = (contents: string, vars: FrontMatterVars) => {
  return contents
    .replace(/^/, toYAML(vars) + '\n\n')
    .replace(/[\r\n]{3,}/g, '\n\n');
};

const toYAML = (vars: FrontMatterVars) => {
  const yaml = `---
${Object.entries(vars)
  .map(
    ([key, value]) =>
      `${key}: ${
        typeof value === 'string' ? `"${value.replace(/"/g, '\\"')}"` : value
      }`,
  )
  .join('\n')}
---`;
  return yaml;
};
