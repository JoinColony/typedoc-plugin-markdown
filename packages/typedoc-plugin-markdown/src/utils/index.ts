export * as frontmatterUtils from './front-matter';
export * from './navigation';
