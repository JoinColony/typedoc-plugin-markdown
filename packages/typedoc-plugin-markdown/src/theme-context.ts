import * as path from 'path';
import { Options, ReferenceType, Reflection } from 'typedoc';
import { MarkdownTheme } from '.';
import { breadcrumbsPartial } from './resources/partials/breadcrumbs';
import { commentPartial } from './resources/partials/comment';
import { commentsPartial } from './resources/partials/comments';
import { declarationMemberPartial } from './resources/partials/declaration-member';
import { declarationTitlePartial } from './resources/partials/declaration-title';
import { groupsPartial } from './resources/partials/groups';
import { hierarchyPartial } from './resources/partials/hierarchy';
import { indexSignatureTitlePartial } from './resources/partials/index-signature-title';
import { memberPartial } from './resources/partials/member';
import { parameterTablePartial } from './resources/partials/parameter-table';
import { propertyTablePartial } from './resources/partials/property-table';
import { referenceMemberPartial } from './resources/partials/reference-member';
import { reflectionPathPartial } from './resources/partials/reflection-path';
import { reflectionTitlePartial } from './resources/partials/reflection-title';
import { signatureMemberPartial } from './resources/partials/signature-member';
import { signatureTitlePartial } from './resources/partials/signature-title';
import { sourcesPartial } from './resources/partials/sources';
import { symbolPartial } from './resources/partials/symbol';
import { tocPartial } from './resources/partials/toc';
import { typePartial } from './resources/partials/type';
import { typeAndParentPartial } from './resources/partials/type-and-parent';
import { typeParameterTablePartial } from './resources/partials/type-parameter-table';
import { indexTemplate } from './resources/templates/index.template';
import { reflectionMemberTemplate } from './resources/templates/member.template';
import { readmeTemplate } from './resources/templates/readme.template';
import { reflectionTemplate } from './resources/templates/reflection.template';
import { URL_PREFIX } from './utils/constants';

export interface TypedocPluginMarkdownOptions {
  hideBreadcrumbs: boolean;
  hideInPageTOC: boolean;
  hidePageTitle: boolean;
  entryDocument: string;
  entryPoints: string[];
  includes: string;
  indexTitle: string;
  media: string;
  namedAnchors: boolean;
  readme: string;
  publicPath: string;
}

function bind<F, L extends any[], R>(fn: (f: F, ...a: L) => R, first: F) {
  return (...r: L) => fn(first, ...r);
}

export class MarkdownThemeContext {
  options: TypedocPluginMarkdownOptions;

  constructor(private theme: MarkdownTheme, options: Options) {
    this.options = options.getRawValues() as TypedocPluginMarkdownOptions;
  }

  relativeURL(url: string | undefined) {
    if (!url) {
      return;
    }
    if (URL_PREFIX.test(url)) {
      return url;
    }
    if (this.options.publicPath) {
      return this.options.publicPath + url;
    }

    const relative = path.relative(
      path.dirname(this.theme.location),
      path.dirname(url),
    );

    return path.join(relative, path.basename(url)).replace(/\\/g, '/');
  }

  urlTo = (reflection: Reflection) => this.relativeURL(reflection.url);

  reflection = () => this.theme.reflection;

  project = () => this.theme.project;

  attemptExternalResolution = (type: ReferenceType) => {
    return this.theme.owner.attemptExternalResolution(type);
  };

  globalsFile = 'modules.md';

  // templates
  indexTemplate = bind(indexTemplate, this);
  readmeTemplate = bind(readmeTemplate, this);
  reflectionTemplate = bind(reflectionTemplate, this);
  reflectionMemberTemplate = bind(reflectionMemberTemplate, this);

  // partials
  breadcrumbsPartial = bind(breadcrumbsPartial, this);
  commentPartial = bind(commentPartial, this);
  commentsPartial = bind(commentsPartial, this);
  declarationMemberPartial = bind(declarationMemberPartial, this);
  signatureMemberPartial = bind(signatureMemberPartial, this);
  memberContainerPartial = bind(memberPartial, this);
  groupsPartial = bind(groupsPartial, this);
  propertyTablePartial = bind(propertyTablePartial, this);
  typeParameterTablePartial = bind(typeParameterTablePartial, this);
  parameterTablePartial = bind(parameterTablePartial, this);
  declarationTitlePartial = bind(declarationTitlePartial, this);
  hierarchyPartial = bind(hierarchyPartial, this);
  indexSignatureTitlePartial = bind(indexSignatureTitlePartial, this);
  symbolPartial = bind(symbolPartial, this);
  reflectionTitlePartial = bind(reflectionTitlePartial, this);
  referenceMemberPartial = bind(referenceMemberPartial, this);
  reflectionPathPartial = bind(reflectionPathPartial, this);
  signatureTitlePartial = bind(signatureTitlePartial, this);
  tocPartial = bind(tocPartial, this);
  typePartial = bind(typePartial, this);
  typeAndParentPartial = bind(typeAndParentPartial, this);
  sourcesPartial = bind(sourcesPartial, this);
}
