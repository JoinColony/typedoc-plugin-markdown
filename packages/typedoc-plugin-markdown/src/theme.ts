import {
  BindOption,
  DeclarationReflection,
  PageEvent,
  ProjectReflection,
  Reflection,
  ReflectionKind,
  Renderer,
  RendererEvent,
  Theme,
  UrlMapping,
} from 'typedoc';
import { MarkdownThemeRenderContext } from './theme-context';
import { getKindPlural } from './utils/helpers';
import { TemplateMapping } from './utils/models';
import { NavigationItem } from './utils/navigation-item';

export class MarkdownTheme extends Theme {
  @BindOption('allReflectionsHaveOwnDocument')
  allReflectionsHaveOwnDocument!: boolean;
  @BindOption('entryDocument')
  entryDocument!: string;
  @BindOption('entryPoints')
  entryPoints!: string[];
  @BindOption('filenameSeparator')
  filenameSeparator!: string;
  @BindOption('readme')
  readme!: string;
  @BindOption('preserveAnchorCasing')
  preserveAnchorCasing!: boolean;

  anchorMap: Record<string, string[]> = {};

  private renderContext?: MarkdownThemeRenderContext;

  static URL_PREFIX = /^(http|ftp)s?:\/\//;

  constructor(renderer: Renderer) {
    super(renderer);

    this.listenTo(this.owner, {
      [RendererEvent.BEGIN]: this.onBeginRenderer,
      [PageEvent.BEGIN]: this.onBeginPage,
    });
  }

  /**
   * Returns the associated render context.
   * @returns
   */
  getRenderContext() {
    if (!this.renderContext) {
      this.renderContext = new MarkdownThemeRenderContext(
        this,
        this.application.options,
      );
    }
    return this.renderContext;
  }

  render(page: PageEvent<Reflection>): string {
    return this.formatContents(page.template(page) as string);
  }

  formatContents(contents: string) {
    return contents;
    /*return (
      contents
        .replace(/[\r\n]{3,}/g, '\n\n')
        .replace(/!spaces/g, '')
        .replace(/^\s+|\s+$/g, '') + '\n'
    );*/
  }

  getUrls(project: ProjectReflection) {
    const urls: UrlMapping[] = [];
    const noReadmeFile = this.readme.endsWith('none');
    if (noReadmeFile) {
      project.url = this.entryDocument;
      urls.push(
        new UrlMapping(this.entryDocument, project, this.reflectionTemplate),
      );
    } else {
      project.url = this.globalsDocument();
      urls.push(
        new UrlMapping(
          this.globalsDocument(),
          project,
          this.reflectionTemplate,
        ),
      );
      urls.push(
        new UrlMapping(this.entryDocument, project, this.readmeTemplate),
      );
    }
    project.children?.forEach((child: Reflection) => {
      if (child instanceof DeclarationReflection) {
        this.buildUrls(child as DeclarationReflection, urls);
      }
    });
    return urls;
  }

  buildUrls(
    reflection: DeclarationReflection,
    urls: UrlMapping[],
  ): UrlMapping[] {
    const mapping = this.mappings.find((mapping) =>
      reflection.kindOf(mapping.kind),
    );
    if (mapping) {
      if (!reflection.url || !MarkdownTheme.URL_PREFIX.test(reflection.url)) {
        const url = this.toUrl(reflection, mapping);
        urls.push(new UrlMapping(url, reflection, mapping.template));
        reflection.url = url;
        reflection.hasOwnDocument = true;
      }

      for (const child of reflection.children || []) {
        if (mapping.isLeaf) {
          this.applyAnchorUrl(child, reflection);
        } else {
          this.buildUrls(child, urls);
        }
      }
    } else if (reflection.parent) {
      this.applyAnchorUrl(reflection, reflection.parent, true);
    }
    return urls;
  }

  toUrl(reflection: DeclarationReflection, mapping: TemplateMapping) {
    return mapping.directory + '/' + this.getUrl(reflection) + '.md';
  }

  getUrl(reflection: Reflection, relative?: Reflection): string {
    let url = reflection.getAlias();

    if (
      reflection.parent &&
      reflection.parent !== relative &&
      !(reflection.parent instanceof ProjectReflection)
    ) {
      url =
        this.getUrl(reflection.parent, relative) + this.filenameSeparator + url;
    }

    return url.replace(/^_/, '');
  }

  applyAnchorUrl(
    reflection: Reflection,
    container: Reflection,
    isSymbol = false,
  ) {
    if (
      container.url &&
      (!reflection.url || !MarkdownTheme.URL_PREFIX.test(reflection.url))
    ) {
      const reflectionId = this.preserveAnchorCasing
        ? reflection.name
        : reflection.name.toLowerCase();

      if (isSymbol) {
        this.anchorMap[container.url]
          ? this.anchorMap[container.url].push(reflectionId)
          : (this.anchorMap[container.url] = [reflectionId]);
      }

      const count = this.anchorMap[container.url]?.filter(
        (id) => id === reflectionId,
      )?.length;

      const anchor = this.toAnchorRef(
        reflectionId + (count > 1 ? '-' + (count - 1).toString() : ''),
      );

      reflection.url = container.url + '#' + anchor;
      reflection.anchor = anchor;
      reflection.hasOwnDocument = false;
    }
    reflection.traverse((child) => {
      if (child instanceof DeclarationReflection) {
        this.applyAnchorUrl(child, container);
      }
    });
  }

  toAnchorRef(reflectionId: string) {
    return reflectionId;
  }

  getNavigation(project: ProjectReflection) {
    const urls = this.getUrls(project);

    const getUrlMapping = (name) => {
      if (!name) {
        return '';
      }
      return urls.find((url) => url.model.name === name);
    };

    const createNavigationItem = (
      title: string,
      url: string | undefined,
      isLabel: boolean,
      children: NavigationItem[] = [],
    ) => {
      const navigationItem = new NavigationItem(title, url);
      navigationItem.isLabel = isLabel;
      navigationItem.children = children;
      const { reflection, parent, ...filteredNavigationItem } = navigationItem;
      return filteredNavigationItem as NavigationItem;
    };
    const navigation = createNavigationItem(project.name, undefined, false);
    const hasReadme = !this.readme.endsWith('none');
    if (hasReadme) {
      navigation.children?.push(
        createNavigationItem('Readme', this.entryDocument, false),
      );
    }
    if (this.entryPoints.length === 1) {
      navigation.children?.push(
        createNavigationItem(
          'Exports',
          hasReadme ? this.globalsDocument() : this.entryDocument,
          false,
        ),
      );
    }
    this.mappings.forEach((mapping) => {
      const kind = mapping.kind[0];
      const items = project.getReflectionsByKind(kind);
      if (items.length > 0) {
        const children = items
          .map((item) =>
            createNavigationItem(
              item.getFullName(),
              (getUrlMapping(item.name) as any)?.url as string,
              true,
            ),
          )
          .sort((a, b) => (a.title > b.title ? 1 : -1));
        const group = createNavigationItem(
          getKindPlural(kind),
          undefined,
          true,
          children,
        );
        navigation.children?.push(group);
      }
    });
    return navigation;
  }

  get mappings() {
    return [
      {
        kind: [ReflectionKind.Module],
        isLeaf: false,
        directory: 'modules',
        template: this.reflectionTemplate,
      },
      {
        kind: [ReflectionKind.Namespace],
        isLeaf: false,
        directory: 'modules',
        template: this.reflectionTemplate,
      },
      {
        kind: [ReflectionKind.Enum],
        isLeaf: false,
        directory: 'enums',
        template: this.reflectionTemplate,
      },
      {
        kind: [ReflectionKind.Class],
        isLeaf: false,
        directory: 'classes',
        template: this.reflectionTemplate,
      },
      {
        kind: [ReflectionKind.Interface],
        isLeaf: false,
        directory: 'interfaces',
        template: this.reflectionTemplate,
      },
      ...(this.allReflectionsHaveOwnDocument
        ? [
            {
              kind: [ReflectionKind.TypeAlias],
              isLeaf: true,
              directory: 'types',
              template: this.memberTemplate,
            },
            {
              kind: [ReflectionKind.Variable],
              isLeaf: true,
              directory: 'variables',
              template: this.memberTemplate,
            },
            {
              kind: [ReflectionKind.Function],
              isLeaf: true,
              directory: 'functions',
              template: this.memberTemplate,
            },
          ]
        : []),
    ];
  }

  protected globalsDocument = () => {
    return this.getRenderContext().globalsDocument;
  };

  protected readmeTemplate = (pageEvent: PageEvent<ProjectReflection>) => {
    return this.getRenderContext().readmeTemplate(pageEvent);
  };

  protected reflectionTemplate = (
    pageEvent: PageEvent<ProjectReflection | DeclarationReflection>,
  ) => {
    return this.getRenderContext().reflectionTemplate(pageEvent);
  };

  protected memberTemplate = (pageEvent: PageEvent<DeclarationReflection>) => {
    return this.getRenderContext().memberTemplate(pageEvent);
  };

  protected onBeginRenderer(event: RendererEvent) {
    this.getRenderContext().project = event.project;
  }

  protected onBeginPage(page: PageEvent) {
    this.getRenderContext().activeLocation = page.url;
    if (page.model instanceof DeclarationReflection) {
      this.getRenderContext().activeReflection = page.model;
    }
  }
}
