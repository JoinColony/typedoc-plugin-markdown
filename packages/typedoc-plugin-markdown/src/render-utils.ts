import attemptExternalResolution from './resources/helpers/attemptExternalResolution';
import breadcrumbsHelper from './resources/helpers/breadcrumbs';
import commentHelper from './resources/helpers/comment';
import commentsHelper from './resources/helpers/comments';
import declarationTitleHelper from './resources/helpers/declaration-title';
import escapeHelper from './resources/helpers/escape';
import hierarchyHelper from './resources/helpers/hierarchy';
import indexSignatureTitleHelper from './resources/helpers/index-signature-title';
import parameterTableHelper from './resources/helpers/parameter-table';
import propertyTableHelper from './resources/helpers/property-table';
import referenceMember from './resources/helpers/reference-member';
import reflectionPathHelper from './resources/helpers/reflection-path';
import reflectionTitleHelper from './resources/helpers/reflection-title';
import relativeUrlHelper from './resources/helpers/relative-url';
import signatureTitleHelper from './resources/helpers/signature-title';
import tocHelper from './resources/helpers/toc';
import typeHelper from './resources/helpers/type';
import typeAndParentHelper from './resources/helpers/type-and-parent';
import typeParameterTableHelper from './resources/helpers/type-parameter-table';
import { MarkdownTheme } from './theme';

export function registerHelpers(theme: MarkdownTheme) {
  attemptExternalResolution(theme);
  breadcrumbsHelper(theme);
  commentHelper(theme);
  commentsHelper();
  declarationTitleHelper(theme);
  escapeHelper();
  hierarchyHelper();
  indexSignatureTitleHelper();
  parameterTableHelper();
  propertyTableHelper();
  referenceMember();
  reflectionPathHelper();
  reflectionTitleHelper(theme);
  relativeUrlHelper(theme);
  signatureTitleHelper(theme);
  tocHelper(theme);
  typeHelper();
  typeAndParentHelper();
  typeParameterTableHelper();
}
