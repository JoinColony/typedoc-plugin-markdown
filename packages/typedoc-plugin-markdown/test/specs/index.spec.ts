import { MarkdownThemeContext } from '../../src';
import { formatContents } from '../../src/utils/format';

import { TestApp } from '../test-app';

describe(`Index:`, () => {
  let testApp: TestApp;
  let context: MarkdownThemeContext;

  beforeAll(async () => {
    testApp = new TestApp(['reflections.ts']);
    await testApp.bootstrap();
    context = testApp.getRenderContext();
    jest.spyOn(context, 'breadcrumbsPartial').mockReturnValue('[breadcrumbs]');
    jest.spyOn(context, 'tocPartial').mockReturnValue('[toc]');
  });

  test(`should compile readme`, () => {
    expect(
      formatContents(
        context.indexTemplate({
          model: testApp.project,
          project: testApp.project,
        } as any),
      ),
    ).toMatchSnapshot();
  });
});
