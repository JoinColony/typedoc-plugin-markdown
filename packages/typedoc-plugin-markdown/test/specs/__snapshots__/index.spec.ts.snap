// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index: should compile readme 1`] = `
"[breadcrumbs]

# Project: typedoc-plugin-markdown

[toc]
"
`;
